src/material/core/private/ripple-loader.ts

```diff
@@ -0,0 +1,168 @@
+ * This service allows us to avoid eagerly creating & attaching MatRipples.
+ * It works by creating & attaching a ripple only when a component is first interacted with.
+ */
+@Injectable({providedIn: 'root'})
+export class MatRippleLoader implements OnDestroy {
+  private _document = inject(DOCUMENT, {optional: true});
+  private _animationMode = inject(ANIMATION_MODULE_TYPE, {optional: true});
+  private _globalRippleOptions = inject(MAT_RIPPLE_GLOBAL_OPTIONS, {optional: true});
+  private _platform = inject(Platform);
+  private _ngZone = inject(NgZone);
+
+  constructor() {
+    this._ngZone.runOutsideAngular(() => {
+      for (const event of rippleInteractionEvents) {
+        this._document?.addEventListener(event, this._onInteraction, eventListenerOptions);
+      }
+    });
+  }
+
+  ngOnDestroy() {
+    for (const event of rippleInteractionEvents) {
+      this._document?.removeEventListener(event, this._onInteraction, eventListenerOptions);
+    }
+  }
+
+  /**
+   * Configures the ripple that will be rendered by the ripple loader.
+   *
+   * Stores the given information about how the ripple should be configured on the host
+   * element so that it can later be retrived & used when the ripple is actually created.
+   */
+  configureRipple(
+    host: HTMLElement,
+    config: {
+      className?: string;
+      centered?: boolean;
+    },
+  ): void {
+    // Indicates that the ripple has not yet been rendered for this component.
+    host.setAttribute(matRippleUninitialized, '');
+
+    // Store the additional class name(s) that should be added to the ripple element.
+    if (config.className || !host.hasAttribute(matRippleClassName)) {
+      host.setAttribute(matRippleClassName, config.className || '');
+    }
+
+    // Store whether the ripple should be centered.
+    if (config.centered) {
+      host.setAttribute(matRippleCentered, '');
+    }
+  }
+
+  /** Returns the ripple instance for the given host element. */
+  getRipple(host: HTMLElement): MatRipple | undefined {
+    if ((host as any).matRipple) {
+      return (host as any).matRipple;
+    }
+    return this.createRipple(host);
+  }
+
+  /** Sets the disabled state on the ripple instance corresponding to the given host element. */
+  setDisabled(host: HTMLElement, disabled: boolean): void {
+    const ripple = (host as any).matRipple as MatRipple | undefined;
+
+    // If the ripple has already been instantiated, just disable it.
+    if (ripple) {
+      ripple.disabled = disabled;
+      return;
+    }
+
+    // Otherwise, set an attribute so we know what the
+    // disabled state should be when the ripple is initialized.
+    if (disabled) {
+      host.setAttribute(matRippleDisabled, '');
+    } else {
+      host.removeAttribute(matRippleDisabled);
+    }
+  }
+
+  /** Handles creating and attaching component internals when a component it is initially interacted with. */
+  private _onInteraction = (event: Event) => {
+    if (event.target === this._document) {
```

Comment:
would it be safer to do `if (!event.target instanceof HTMLElement)`?

Reply:
I think @crisbeto pointed out that this causes problem for ssr? I'll try to find the original comment

Reply:
Found it
https://github.com/angular/components/pull/26568#discussion_r1151469748